substrate/frame/support/procedural/src/construct_runtime/expand/inherent.rs

```diff
@@ -204,13 +207,59 @@ pub fn expand_outer_inherent(
 			}
 		}
 
+		impl #scrate::traits::EnsureInherentsAreOrdered<#block> for #runtime {
+			fn ensure_inherents_are_ordered(block: &#block, num_inherents: usize) -> Result<(), #scrate::__private::sp_inherents::InherentOrderError> {
+				use #scrate::inherent::ProvideInherent;
+				use #scrate::traits::{IsSubType, ExtrinsicCall};
+				use #scrate::sp_runtime::traits::Block as _;
+				use #scrate::__private::sp_inherents::{InherentOrder, InherentOrderError};
+
+				// The last order that we encountered. Must increase monotonic.
+				let mut last: Option<InherentOrder> = None;
+
+				for (i, xt) in block.extrinsics().iter().take(num_inherents).enumerate() {
+					let call = <#unchecked_extrinsic as ExtrinsicCall>::call(xt);
+
+					let mut current: Option<InherentOrder> = None;
+
+					#(
+						if let Some(call) = IsSubType::<_>::is_sub_type(call) {
+							if #pallet_names::is_inherent(&call) {
+								let order = #pallet_names::inherent_order().unwrap_or(InherentOrder::Index(#pallet_indices as u32));
```

Comment:
Inherent order shouldn't depend on pallet indexes, because pallet indexes can't be modified for a chain that's already been in production for a long time (this breaks metadata and SCALE encoding). I suggest instead defaulting to the order of declaration in the construct_runtime macro, and allowing the runtime to define a different explicit order.